### DISCUSSION TITLE
Consider moving to a different forge

### DISCUSSION THREAD

**Discussion by josealberto4444** [axelboc/anki-ultimate-geography#612] [General]
Hi,

With Github becoming more and more hostile with time since its adquisition by Microsoft, should we consider moving to a different forge? I think [Codeberg](https://codeberg.org/) is the best option currently, because [sourcehut](https://sr.ht/) might scare occasional contributors.

I know I'm not the most regular contributor here, but I just wanted you to consider (or, at least, to know) other options. =)

Thank you! ^^

**Comment by aplaice** [comment]
Thanks for bringing this up!

I'm not too happy about using GitHub, and would very much prefer a FOSS forge, but I think that given the desired low technical bar for contributions, we should be particularly wary of migrating off the lowest-common denominator. 

OTOH given that (from what I very vaguely remember) many of our contributors (as non-programmers) were first-time users of GitHub, the network effect might not actually be _that_ important a consideration for us, which would suggest that while we likely need a higher-than-average bar for ease-of-use, we might be less tied to GitHub in particular than the average project. (i.e. sourcehut is likely still off the table, but Codeberg maybe not).

However, I don't really see us migrating fully off GitHub, and only having a mirror (likely with GitHub still as the canonical repo :( ).

<hr/>

There's also the matter of issues and pull requests that contain a very large amount of discussions, and which we liberally link to. [Codeberg allows](https://codeberg.org/Recommendations/Mirror_to_Codeberg) [migrating them](https://docs.codeberg.org/advanced/migrating-repos/), but AFAIU does not have an option for continually mirroring them (bidirectionally or even unidirectionally).

If there were an option to avoid numbering clashes — having different issues/pull requests with the same number on different forges — for instance by forcing the non-GitHub forge to start its own numbering from, say, 10000 (?), then an initial migration and then having separate, but non-colliding issues/PRs on the two forges (with the hope that eventually™ they might be "merged"), might perhaps be OK, though.

**Reply by josealberto4444** [reply]
> There's also the matter of issues and pull requests that contain a very large amount of discussions, and which we liberally link to. Codeberg allows migrating them, but AFAIU does not have an option for continually mirroring them (bidirectionally or even unidirectionally).

There is work going on in forgefed, but I don't think Github would use it, anyway.

I understand the reasons and I agree. I wanted to bring the discussion to see your point of view (and of the rest of maintainers yet). ^^

Thank you!

**Reply by aplaice** [reply]
> (and of the rest of maintainers yet). ^^

Yes, I would obviously not make any decisions about this without my co-maintainers' approval!

**Comment by axelboc** [comment]
> With Github becoming more and more hostile with time since its acquisition by Microsoft

Could you please just expand a bit on that? I haven't heard/experienced anything like that so far, but I don't pay much attention to tech news, so I've probably missed something... Follow-up question: does this hostility affect public orgs, like UG, or our existing/future contributors in any way?

> There's also the matter of issues and pull requests that contain a very large amount of discussions, and which we liberally link to.

Big point for me. We've often referred to such discussions in the past. There are links here and there, including in `CONTRIBUTING`, to issues, PRs and even directly to specific comments within those, so we really can't lose all this history. Let's not forget also the release notes (the text can be migrated to a `CHANGELOG` but not the assets) and the couple of wiki pages.

**Reply by jeprecated** [reply]
Seconded, I have not heard much about Github being hostile. Interested to know more though!

**Reply by josealberto4444** [reply]
> Could you please just expand a bit on that? I haven't heard/experienced anything like that so far, but I don't pay much attention to tech news, so I've probably missed something...

The most important changes, in my opinion, are:

- Blocking developers in Iran, Crimea, Cuba, North Korea, and Syria (this happened some years ago, and I think it was reverted, at least partially). Could be happening now with Palestine, but I don't know.
- Copilot training.
- Github CLI (EEE strategy).

But the most annoying for me are:

- Github discussions are not accesible from third party clients (EEE strategy again).
- SSH clone link is not shown in any repo if you haven't logged in.
- Developers are now forced to use 2FA.
- Web interface is getting slower as years go by.

The first two force me to only use Github in my computer (octodroid is now almost useless as a significant amount of the activity is in discussions) and having to log in every time (as I clean cookies regularly) even if I only want to read the notifications (that was the reason my activity in this repo dropped to almost nothing when discussions appeared). Due to the third, I had to install yet another app on my phone, so I had to make a bit of research to find a good FLOSS app.

> Follow-up question: does this hostility affect public orgs, like UG, or our existing/future contributors in any way?

Well, that depends on the country of the maintainers or future maintainers, or if we see some bits of the code of this repo as part of Copilot output in another place, and also on the future choices of Microsoft. At least, the annoying bits affect me. I think all these reasons (and probably some more that I didn't find at the time of writing this) affect more users as well, as there have already been some waves of people leaving Github and switching to other forges.

> Big point for me. We've often referred to such discussions in the past. There are links here and there, including in CONTRIBUTING, to issues, PRs and even directly to specific comments within those, so we really can't lose all this history. Let's not forget also the release notes (the text can be migrated to a CHANGELOG but not the assets) and the couple of wiki pages.

That depends on the quality of the import feature of Codeberg, and may require manual intervention in the migration, but I don't think it would be lost. If we considered switching in the future, we could ask them (even for help), they are very nice and helpful people.

**Reply by axelboc** [reply]
> Blocking developers in Iran, Crimea, Cuba, North Korea, and Syria (this happened some years ago, and I think it was reverted, at least partially). Could be happening now with Palestine, but I don't know.

Big downer indeed. This alone is enough to convince me.

> Github discussions are not accesible from third party clients (EEE strategy again).

This too, for sure. Sorry to hear you've had trouble contributing since we enabled the Discussions feature on the repo. :confounded: 

> Developers are now forced to use 2FA.

This, however, I think is a good thing! :sweat_smile: The new forge would have to at least enforce 2FA for maintainers.

**Reply by josealberto4444** [reply]
> Sorry to hear you've had trouble contributing since we enabled the Discussions feature on the repo. 😖

Well, it just stopped me from checking notifications from my mobile, which I used to keep in touch with the project, but I wasn't at my best time and maybe I couldn't contribute too much anyway, so don't worry. I don't want any of you to feel bad at all. <3

Thank you all for the understanding and the discussion. I feel this project like a safe place to bring up these type of discussions because you are all always very kind and understanding, and it's a pleasure to talk to you and receive your feedback. ^^
